### PR TITLE
2018 GMC Acadia Denali Support

### PR DESCRIPTION
**Pull request by CryptoKylan**
added 2018 GMC Acadia Denali fingerprint to volt as requested. lat and long in OP work however still a bit buggy

### CODE REVIEW

**Comment by rbiasini** [comment]
@CryptoKylan , great!
What do you think it's still buggy?

File: selfdrive/car/gm/carcontroller.py [hunk 1]
```diff
@@ -14,6 +14,11 @@ def __init__(self, car_fingerprint):
       self.STEER_STEP = 2              # how often we update the steer cmd
       self.STEER_DELTA_UP = 7          # ~0.75s time to peak torque (255/50hz/0.75s)
       self.STEER_DELTA_DOWN = 17       # ~0.3s from peak torque to zero
+    elif car_fingerprint == CAR.ACADIA_DENALI:
```
Comment (rbiasini) [inline selfdrive/car/gm/carcontroller.py]: @CryptoKylan , why duplicating the code?
if car_fingerprint is in [CAR.VOLT, CAR.ACADIA_DENALI] should be ok, right?

File: selfdrive/car/gm/carcontroller.py [hunk 2]
```diff
@@ -115,6 +120,9 @@ def update(self, sendcan, enabled, CS, frame, actuators, \
       if self.car_fingerprint == CAR.VOLT:
         can_sends.append(gmcan.create_steering_control(self.packer_pt,
           canbus.powertrain, apply_steer, idx, lkas_enabled))
+      if self.car_fingerprint == CAR.ACADIA_DENALI:
```
Comment (rbiasini) [inline selfdrive/car/gm/carcontroller.py]: same as above

Comment (CryptoKylan) [reply selfdrive/car/gm/carcontroller.py]: i duplicated alot of the settings as a starting point. im really not sure what will be different from the volt but i know some settings will need to be different from the volt.

 i am having an issue with steering temp unavailable message after different amounts of time.  im wondering if it has to do with the volt being electric? maybe something to do with regen breaking throwing errors? im still very new to openpilot so any help would be greatly appreciated

https://community.comma.ai/cabana/?route=7cc2a8365b4dd8a9%7C2018-08-28--19-07-07&max=6&url=https%3A%2F%2Fchffrprivate.blob.core.windows.net%2Fchffrprivate3%2Fv2%2F7cc2a8365b4dd8a9%2F2d522ddc11f1292d5dd307d63aa0321b_2018-08-28--19-07-07

**Comment by CryptoKylan** [comment]
i am having an issue with steering temp unavailable message after different amounts of time. im wondering if it has to do with the volt being electric? maybe something to do with regen breaking throwing errors? im still very new to openpilot so any help would be greatly appreciated

https://community.comma.ai/cabana/?route=7cc2a8365b4dd8a9%7C2018-08-28--19-07-07&max=6&url=https%3A%2F%2Fchffrprivate.blob.core.windows.net%2Fchffrprivate3%2Fv2%2F7cc2a8365b4dd8a9%2F2d522ddc11f1292d5dd307d63aa0321b_2018-08-28--19-07-07

**Comment by CryptoKylan** [comment]
im just adding some new drives in hopes someone can help figure out why im loosing steering.  OP lasted 20 mins straight without issue on this drive. steering cuts out at about 1700 secs. LKATorqueDeliveredStatus goes to 3 and ASCMLKASteeringCmdActive becomes 0.

the whole 20 mins OP operated as it should. stop and go no problem, adjusting the speed up and down dramatically quite a bit without issue, changed lanes multiple times. taking sharp turns without needing any help from me. only a couple times i had to apply a small amount of force on the wheel to help complete the turn. i engaged and disengaged a few times before the long run also.

just to be clear when this happens OP ACC will stay engaged, i just loose steering. in fact i can remove the no.entry flag from OP so i at least can continue to use ACC as i do alot of long driving and id rather have that than nothing at all in the mean time.

https://community.comma.ai/cabana/?route=7cc2a8365b4dd8a9%7C2018-09-02--19-43-52&max=30&url=https%3A%2F%2Fchffrprivate.blob.core.windows.net%2Fchffrprivate3%2Fv2%2F7cc2a8365b4dd8a9%2Fb9a4cb5937d394050e60b9f509e33329_2018-09-02--19-43-52

i pulled over and restarted the car and got another long run.  this time LKATorqueDeliveredStatus went to 3 in between me disengaging OP and trying to engage again

888 secs in i disengage and about 930 secs LKATorqueDeliveredStatus goes to 3 again while OP has not been engaged again.
https://community.comma.ai/cabana/?route=7cc2a8365b4dd8a9%7C2018-09-02--20-45-35&max=40&url=https%3A%2F%2Fchffrprivate.blob.core.windows.net%2Fchffrprivate3%2Fv2%2F7cc2a8365b4dd8a9%2F4bee816971fc4641ba0535e2c1d9b35a_2018-09-02--20-45-35

i dont know what else to try.

**Comment by CryptoKylan** [comment]
this is working now.  what else do you need from me to push this through?

almost 2 hours straight with no disengagements 

https://community.comma.ai/cabana/?route=7cc2a8365b4dd8a9%7C2018-09-16--22-30-45&max=114&url=https%3A%2F%2Fchffrprivate.blob.core.windows.net%2Fchffrprivate3%2Fv2%2F7cc2a8365b4dd8a9%2F901c96f7d6ff89a81285c595a62b59b1_2018-09-16--22-30-45


another almost hour

https://community.comma.ai/cabana/?route=7cc2a8365b4dd8a9%7C2018-09-16--22-30-45&max=114&url=https%3A%2F%2Fchffrprivate.blob.core.windows.net%2Fchffrprivate3%2Fv2%2F7cc2a8365b4dd8a9%2F901c96f7d6ff89a81285c595a62b59b1_2018-09-16--22-30-45

File: selfdrive/car/gm/interface.py [hunk 3]
```diff
@@ -88,12 +88,23 @@ def get_params(candidate, fingerprint):
...
+      ret.steerRatio = 16
+      ret.steerRatioRear = 0.
+      ret.centerToFront = ret.wheelbase * 0.4
 
     elif candidate == CAR.CADILLAC_CT6:
       # engage speed is decided by pcm
       ret.minEnableSpeed = -1
       # kg of standard extra cargo to count for drive, gas, etc...
-      ret.mass = 4016. * CV.LB_TO_KG + std_cargo
+      ret.mass = 4016 * CV.LB_TO_KG + std_cargo
```
Comment (vntarasov) [inline selfdrive/car/gm/interface.py]: CarParams.mass is Float32, why change? Same for CAR.ACADIA_DENALI

File: selfdrive/car/gm/interface.py [hunk 4]
```diff
@@ -88,12 +88,23 @@ def get_params(candidate, fingerprint):
       ret.steerRatio = 15.7
       ret.steerRatioRear = 0.
       ret.centerToFront = ret.wheelbase * 0.4 # wild guess
+    
+    elif candidate == CAR.ACADIA_DENALI:
+      # supports stop and go, but initial engage must be above 18mph (which include conservatism)
```
Comment (vntarasov) [inline selfdrive/car/gm/interface.py]: Comment doesn't match minEnableSpeed of -1.

File: selfdrive/car/gm/carcontroller.py [hunk 5]
```diff
@@ -15,6 +15,11 @@ def __init__(self, car_fingerprint):
       self.STEER_STEP = 2              # how often we update the steer cmd
       self.STEER_DELTA_UP = 7          # ~0.75s time to peak torque (255/50hz/0.75s)
       self.STEER_DELTA_DOWN = 17       # ~0.3s from peak torque to zero
+    elif car_fingerprint == CAR.ACADIA_DENALI:
+      self.STEER_MAX = 255
+      self.STEER_STEP = 3              # how often we update the steer cmd
```
Comment (vntarasov) [inline selfdrive/car/gm/carcontroller.py]: @rbiasini, @CryptoKylan says stock LKA is at 50 Hz, but with STEER_STEP = 2 it sometimes faults out (hence STEER_STEP = 3).

@CryptoKylan, have you experimented with smaller DELTA_UP but still "STEER_STEP = 2"? What's the highest "delta up" you have seen in stock LKA logs?

File: selfdrive/car/gm/carcontroller.py [hunk 6]
```diff
@@ -15,6 +15,11 @@ def __init__(self, car_fingerprint):
...
+      self.STEER_DELTA_UP = 5          # ~0.75s time to peak torque (255/50hz/0.75s)
```
Comment (vntarasov) [inline selfdrive/car/gm/carcontroller.py]: Comment doesn't match current STEER_STEP

**Comment by energee** [comment]
It should be letting the message through, this code is restricting the panda from sending torque in situations where "violations" occur.

**Comment by CryptoKylan** [comment]
right

> It should be letting the message through, this code is restricting the panda from sending torque in situations where "violations" occur.

right.  thats why im wondering why the code first resets and sends zero torque to just stop the msg entirely in the next step

// reset to 0 if either controls is not allowed or there's a violation
    if (violation || !current_controls_allowed) {
      gm_desired_torque_last = 0;
      gm_rt_torque_last = 0;
      gm_ts_last = ts;
    }

    if (violation) {
      return false;

it sort of seems redundant doesn't it?

wouldn't it make more sense to just send zero/reset torque and let the msg through so it doesnt cause an EPS fault when the car sees a missing message on disengagement?

**Pull request closed without merging**